src/ir/possible-contents.cpp

```diff
@@ -0,0 +1,1628 @@
+    // in the reference, so mark us as the parent of the ref, and we will
+    // handle all of this in a special way during the flow. Note that we do
+    // not even create a DataLocation here; anything that we need will be
+    // added during the flow.
+    addSpecialChildParentLink(curr->ref, curr);
+  }
+  void visitStructSet(StructSet* curr) {
+    if (curr->ref->type == Type::unreachable) {
+      return;
+    }
+    // See comment on visitStructGet. Here we also connect the value.
+    addSpecialChildParentLink(curr->ref, curr);
+    addSpecialChildParentLink(curr->value, curr);
+  }
+  // Array operations access the array's location, parallel to how structs work.
+  void visitArrayGet(ArrayGet* curr) {
+    if (!isRelevant(curr->ref)) {
+      addRoot(curr);
+      return;
+    }
+    addSpecialChildParentLink(curr->ref, curr);
+  }
+  void visitArraySet(ArraySet* curr) {
+    if (curr->ref->type == Type::unreachable) {
+      return;
+    }
+    addSpecialChildParentLink(curr->ref, curr);
+    addSpecialChildParentLink(curr->value, curr);
+  }
+
+  void visitArrayLen(ArrayLen* curr) {
+    // TODO: optimize when possible (perhaps we can infer a Literal for the
+    //       length)
+    addRoot(curr);
+  }
+  void visitArrayCopy(ArrayCopy* curr) {
+    if (curr->type == Type::unreachable) {
+      return;
+    }
+    // Our flow handling of GC data is not simple: we have special code for each
+    // read and write instruction. Therefore, to avoid adding special code for
+    // ArrayCopy, model it as a combination of an ArrayRead and ArrayWrite, by
+    // just emitting fake expressions for those. The fake expressions are not
+    // part of the main IR, which is potentially confusing during debugging,
+    // however, which is a downside.
+    Builder builder(*getModule());
+    auto* get = builder.makeArrayGet(curr->srcRef, curr->srcIndex);
+    visitArrayGet(get);
+    auto* set = builder.makeArraySet(curr->destRef, curr->destIndex, get);
+    visitArraySet(set);
+  }
+
+  // TODO: Model which throws can go to which catches. For now, anything thrown
+  //       is sent to the location of that tag, and any catch of that tag can
+  //       read them.
+  void visitTry(Try* curr) {
+    receiveChildValue(curr->body, curr);
+    for (auto* catchBody : curr->catchBodies) {
+      receiveChildValue(catchBody, curr);
+    }
+
+    auto numTags = curr->catchTags.size();
+    for (Index tagIndex = 0; tagIndex < numTags; tagIndex++) {
+      auto tag = curr->catchTags[tagIndex];
+      auto* body = curr->catchBodies[tagIndex];
+
+      auto params = getModule()->getTag(tag)->sig.params;
+      if (params.size() == 0) {
+        return;
+      }
+
+      // Find the pop of the tag's contents. The body must start with such a
+      // pop, which might be of a tuple.
+      auto pops = EHUtils::findPops(body);
+      assert(pops.size() == 1);
+      auto* pop = pops[0];
```

Comment:
```suggestion
      auto* pop = EHUtils::findPop(body);
      if (!pop) {
        continue;
      }
```

Reply:
Oh, sorry, I did do this change to use `findPop` and not `Pops` but looks like I didn't push it yesterday. But it was pushed today among the other updates.

Good point about the case of null, I didn't handle that, but I'm adding a commit now. I think we can assert on `pop` existing since we exited earlier if the tag had no params, is that correct?

Reply:
Ah right, we have an early exit in case of no params, in which case we don't need this `if`... By the way is it ok to return? Shouldn't we continue?